File: example/modules/perf-lab/ios/ProbeBridge.swift
import Foundation

/// Talks to `NitroMapsPerfProbeBridge` inside react-native-better-maps by
/// selector, so this module never imports the library's Swift module (which
/// is compiled with C++ interop). Every selector is argument-free.
enum ProbeBridge {
  private static var bridgeClass: AnyObject? {
    NSClassFromString("NitroMapsPerfProbeBridge")
  }

  private static func callObject(_ name: String) -> AnyObject? {
    guard let bridge = bridgeClass else {
      return nil
    }
    let selector = NSSelectorFromString(name)
    guard bridge.responds(to: selector) else {
      return nil
    }
    return bridge.perform(selector)?.takeUnretainedValue()
  }

  private static func callVoid(_ name: String) {
    guard let bridge = bridgeClass else {
      return
    }
    let selector = NSSelectorFromString(name)
    guard bridge.responds(to: selector) else {
      return
    }
    _ = bridge.perform(selector)
  }

  static var isAvailable: Bool {
    (callObject("probesAvailable") as? NSNumber)?.boolValue ?? false
  }

  static func setEnabled(_ enabled: Bool) {
    callVoid(enabled ? "enableProbes" : "disableProbes")
  }

  static func drainJSON() -> String {
    (callObject("drainJSON") as? String) ?? "{\"spans\":[],\"dropped\":0}"
  }
}

File: example/modules/perf-lab/android/src/main/java/expo/modules/perflab/ProbeBridge.kt
package expo.modules.perflab

/**
 * Reaches `NitroMapsPerfProbeBridge` inside react-native-better-maps through
 * reflection so this module does not depend on the library's Gradle module.
 */
internal object ProbeBridge {
  private const val CLASS_NAME = "com.margelo.nitro.nitromaps.NitroMapsPerfProbeBridge"

  private val bridge: Class<*>? = try {
    Class.forName(CLASS_NAME)
  } catch (_: ClassNotFoundException) {
    null
  }

  fun isAvailable(): Boolean =
    bridge?.getMethod("isAvailable")?.invoke(null) as? Boolean ?: false

  fun setEnabled(enabled: Boolean) {
    bridge?.getMethod("setEnabled", Boolean::class.javaPrimitiveType)?.invoke(null, enabled)
  }

  fun drainJson(): String =
    bridge?.getMethod("drainJson")?.invoke(null) as? String ?: "{\"spans\":[],\"dropped\":0}"
}
